chain/beacon/chain.go

```diff
@@ -217,15 +219,31 @@ func (c *chainStore) shouldSync(last *chain.Beacon, newB likeBeacon) bool {
 	return newB.GetRound() > last.GetRound()+1
 }
 
-// RunSync will sync up with other nodes and fill the store. If upTo is equal to
-// 0, then it will follow the chain indefinitely. If peers is nil, it uses the
-// peers of the current group.
+// RunSync will sync up with other nodes and fill the store.
+// It will start from the latest stored beacon. If upTo is equal to 0, then it
+// will follow the chain indefinitely. If peers is nil, it uses the peers of
+// the current group.
 func (c *chainStore) RunSync(upTo uint64, peers []net.Peer) {
-	if peers == nil {
+	if len(peers) == 0 {
 		peers = toPeers(c.crypto.GetGroup().Nodes)
 	}
 
-	c.syncm.RequestSync(peers, upTo)
+	c.syncm.RequestSync(upTo, peers)
+}
+
+// RunReSync will sync up with other nodes to repair the invalid beacons in the store.
+func (c *chainStore) RunReSync(ctx context.Context, faultyBeacons []uint64, peers []net.Peer, cb func(r, u uint64)) error {
+	// we do this check here because the SyncManager doesn't have the notion of group
+	if len(peers) == 0 {
+		peers = toPeers(c.crypto.GetGroup().Nodes)
+	}
+
+	return c.syncm.CorrectPastBeacons(ctx, faultyBeacons, peers, cb)
+}
+
+// ValidateChain will validate the chain in the chain store up to the given beacon.
+func (c *chainStore) ValidateChain(ctx context.Context, upTo uint64, cb func(r, u uint64)) ([]uint64, error) {
```

Comment:
Can you specify what ValidateChain means here ? What's different than RunResync ? (I think I know but just looking at comment here it's just repeating the name of the function which is not helping much).